File: leetcode/208_Implement_Trie_Prefix_Tree.py
import unittest

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        node = self.root
        for idx in (ord(char) - 97 for char in word):
            if node.alp[idx] is None:
                node.alp[idx] = Node()
            node = node.alp[idx]
        node.is_word = True

    def search(self, word: str) -> bool:
        node = self.root
        for idx in (ord(char) - 97 for char in word):
            if node.alp[idx] is None:
                return False
            node = node.alp[idx]
        return node.is_word

    def startsWith(self, prefix: str) -> bool:
        node = self.root
        for idx in (ord(char) - 97 for char in prefix):
            if node.alp[idx] is None:
                return False
            node = node.alp[idx]
        return True

class Node: 
    def __init__(self):
        self.alp = [None] * 26
        self.is_word = False

class TestTrie(unittest.TestCase):
    def test_1(self) -> None:
        trie = Trie()
        trie.insert("apple")
        self.assertTrue(trie.search("apple"))
        self.assertFalse(trie.search("app"))
        self.assertTrue(trie.startsWith("app"))
        trie.insert("app")
        self.assertTrue(trie.search("app"))
        

File: leetcode/208_Implement_Trie_Prefix_Tree.rs

struct Node {
    children: [Option<Box<Node>>; Trie::CHAR_N],
    terminal: bool,
}

impl Node {
    fn new() -> Self {
        const INIT: Option<Box<Node>> = None;
        Self { children: [INIT; Trie::CHAR_N], terminal: false}
    }
}

struct Trie {
    root: Box<Node>,
}

impl Trie {

    const CHAR_N: usize = 26;
    const CHAR_SHIFT: u8 = b'a';

    fn new() -> Self {
        Self { root: Box::new(Node::new()) }
    }
    
    fn insert(&mut self, word: String) {
        let mut node = &mut self.root;
        for ch in word.as_bytes().iter().map(|v| v - Trie::CHAR_SHIFT) {
            if node.children[ch as usize].is_none() {
                node.children[ch as usize] = Some(Box::new(Node::new()));
            }
            node = node.children[ch as usize].as_mut().unwrap();
        }
        node.terminal = true;
    }
    
    fn search(&self, word: String) -> bool {
        let mut node = &self.root;
        for ch in word.as_bytes().iter().map(|v| v - Trie::CHAR_SHIFT) {
            match &node.children[ch as usize] {
                Some(child) => { node = child },
                None => { return false}
            } 
        }
        node.terminal
    }
    
    fn starts_with(&self, prefix: String) -> bool {
        let mut node = &self.root;
        for ch in prefix.as_bytes().iter().map(|v| v - Trie::CHAR_SHIFT) {
            match &node.children[ch as usize] {
                Some(child) => { node = child },
                None => { return false }
            } 
        }
        true        
    }
}

#[test]
fn test_1() {
    let mut trie = Trie::new();
    trie.insert("apple".to_string());
    assert!(trie.search("apple".to_string()));
    assert!(!trie.search("app".to_string()));
    assert!(trie.starts_with("app".to_string()));
    trie.insert("app".to_string());
    assert!(trie.search("app".to_string()));
}    

